sql: Disallow index span constraints from mixed type comparison ops

nvb opened the pull request:
Fixes #4313.

Previously we allowed the use of mixed-type comparison operations
to be used to constrain index spans. This resulted in incorrect
constraints because the encoding of the constraint was for the wrong
type. This (temporary?) fix removes these invalid index constraints.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5175)

<!-- Reviewable:end -->

petermattis commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/select.go, line 1184 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5175#-KCcLut2o2zwwkSjg-ZZ:-KCcLut2o2zwwkSjg-Z_:-2093701000) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ccacb3b9cab21cc2f590681d6624069b304be2a/sql/select.go#L1184)):*
Curious why you decided to add `ComparisonExpr.MixedType` vs adding a method. Seems a bit awkward to have to forward the `MixedType` state as you do above, and I'm not confident that is being done correctly through normalization and simplification.

Another thought: rather that disallowing these comparison expressions, perhaps `analyzeExpr` should normalize them (when possible) so that the comparison is on the same types. If you do a comparison like `a > 1.5` and `a` is an `INT` we could normalize/simplify that to `a > 1`. Granted, this would be a lot more work than what you've done here. I think a stop-gap solution is appropriate in the beta timeframe. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5175)*
<!-- Sent from Reviewable.io -->

tamird commented:
I thought we were going to go the route of disallowing mixed-type comparisons (and upcasting expressions in certain circumstances instead). Have we decided against that?

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5175#-:-KCcQ3W8RxseNT450XoM:606985396)*
<!-- Sent from Reviewable.io -->

nvb committed:
sql: Disallow index span constraints from mixed type comparison ops

Fixes #4313.

Previously we allowed the use of mixed-type comparison operations
to be used to constrain index spans. This resulted in incorrect
constraints because the encoding of the constraint was for the wrong
type. This (temporary?) fix removes these invalid index constraints.

nvanbenschoten commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/select.go, line 1184 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5175#-KCcLut2o2zwwkSjg-ZZ:-KCqHAKECQKQeX-_L-Jz:441667832) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ccacb3b9cab21cc2f590681d6624069b304be2a/sql/select.go#L1184)):*
I had wanted to try to avoid another traversal of the select filter because I figured these were pretty expensive and could take an arbitrary amount of time, especially if we were just type checking again. What I forget to take into consideration was all of the simplification we do during the index selection phase. Since the disjoint `and` expressions used for constraint selection are flattened, this should be quick to determine during constraint population. I've reworked the strategy so PTAL. 

The issue I have with the second idea is that it introduces a lot of extra logic to deal with conservative type conversions. It would get pretty tricky with statements like `NOT(a > 1.5)`. I guess this would be manageable after our flattening step in simplification, so it might be worth investing some time into this later. However, with a new type system coming in soon that might obviate the need for this, I'd rather wait and go with the quicker solution for now.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5175)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/select.go, line 1184 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5175#-KCcLut2o2zwwkSjg-ZZ:-KCqKvwmT4e9MgT5EP4q:-551752265) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ccacb3b9cab21cc2f590681d6624069b304be2a/sql/select.go#L1184)):*
This looks a lot better. I agree that my second idea isn't appropriate right now. But please file an issue about it as a future enhancement. 

---

*[sql/testdata/select_index, line 247 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5175#-KCqLNBmCK9bVsfnvcWw:-KCqLNBnuABx80_PVSVt:-664202728) ([raw file](https://github.com/cockroachdb/cockroach/blob/4afa3309041857fb62a4a310cd93bcf0947d9994/sql/testdata/select_index#L247)):*
Are we handling mixed-type simplifications correctly? Does something like: `SELECT b FROM t WHERE a = 1 AND a = 1.0` work? I'm worried that we'll panic a comparison on `1` and `1.0`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5175)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5175#-:-KCqLjBLehORLEqcdpPD:1417088721)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[sql/select.go, line 1184 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5175#-KCcLut2o2zwwkSjg-ZZ:-KCqPhISBfYcsoUlXXo3:-17236123) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ccacb3b9cab21cc2f590681d6624069b304be2a/sql/select.go#L1184)):*
Done. #5235.

---

*[sql/testdata/select_index, line 247 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5175#-KCqLNBmCK9bVsfnvcWw:-KCqP5j5DeZvfAAEOzYX:-1572425906) ([raw file](https://github.com/cockroachdb/cockroach/blob/4afa3309041857fb62a4a310cd93bcf0947d9994/sql/testdata/select_index#L247)):*
I have a fix I'm about to push for https://github.com/cockroachdb/cockroach/issues/3633 that should take care of that.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5175)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[sql/testdata/select_index, line 247 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5175#-KCqLNBmCK9bVsfnvcWw:-KCq_rOELUysSnmPZH5X:809632866) ([raw file](https://github.com/cockroachdb/cockroach/blob/4afa3309041857fb62a4a310cd93bcf0947d9994/sql/testdata/select_index#L247)):*
It looks like a simplification like that won't panic, but it will just have an incorrect result. See #5238.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5175)*
<!-- Sent from Reviewable.io -->
